M2/Macaulay2/packages/Msolve.m2

```diff
@@ -0,0 +1,548 @@
+    callStr:=msolveEXEwithOpts(opt)|" -g 2 -f "|mIn|" -o "|mOut;
+    run(callStr);
+    if opt#Verbosity>0 then<<"done msolve"<<endl;
+    msolGB:=readMsolveOutputFile(R, mOut);
+    return gens forceGB msolGB;
+    );
+msolveLeadMonomials=method(TypicalValue=>Matrix, Options=>{Verbosity=>0, "number of threads"=>allowableThreads});
+msolveLeadMonomials(Ideal):=opt->(I)->(
+    if not inputOkay(I) then error "Problem with input, please refer to documentation.";
+    mIn:=temporaryFileName()|".ms";
+    R:=ring I;
+    kk:=coefficientRing(R);
+    gR:=toString(gens R);
+    l1:=substring(1,#gR-2,gR);
+    l2:=char R;
+    gI:=toString flatten entries gens I;
+    if (isField(kk) and (char(kk)==0)) then gI=replace("[)(]","",gI);
+    Igens:=replace(" ",""|newline,substring(1,#gI-2,gI));
+    inStr:=l1|newline|l2|newline|Igens;
+    mIn<<inStr<<close;
+    mOut:=temporaryFileName()|".ms";
+    if opt#Verbosity>0 then<< "msolve input file is called: " << mIn << endl;
+    if opt#Verbosity>0 then<< "msolve output file is called: " << mOut << endl;
+    callStr:=msolveEXEwithOpts(opt)|" -g 1 -f "|mIn|" -o "|mOut;
+    run(callStr);
+    msolGB:=readMsolveOutputFile(R, mOut);
+    return gens forceGB msolGB;
+    );
+msolveEliminate=method(Options=>{Verbosity=>0, "number of threads"=>allowableThreads});
+msolveEliminate(Ideal,RingElement):=Ideal =>opt-> (I,elimvar)->(
+    return msolveEliminate(I,{elimvar});
+    );
+msolveEliminate(RingElement,Ideal):=Ideal => opt->(elimvar,I) ->
+    msolveEliminate(I,{elimvar})
+msolveEliminate(List,Ideal):=Ideal => opt->(elimvars,I) -> 
+    msolveEliminate(I,elimvars)
+msolveEliminate(Ideal,List):=Ideal => opt->(J,elimvars)->(
+    if not inputOkay(J) then error "Problem with input, please refer to documentation.";
+    mIn:=temporaryFileName()|".ms";
+    S:=ring J;
+    kk:=coefficientRing(S);
+    keepvars:=select(gens S, s->not member(s,elimvars)); 
+    elimNum:=length(elimvars);
+    --R:=kk(monoid [join(elimvars,keepVars),MonomialOrder=>{numgens(S):1}]);
+    R:=kk(monoid [join(elimvars,keepvars),MonomialOrder=>{#elimvars, #keepvars}]);
+    elimR:=kk(monoid [keepvars]);
+    I:=sub(J,R);
+    gR:=toString(gens R);
+    l1:=substring(1,#gR-2,gR);  -- removing parentheses
+    l2:=char R;
+    gI:=toString flatten entries gens I;
+    if (isField(kk) and (char(kk)==0)) then gI=replace("[)(]","",gI);
+    Igens:=replace(" ",""|newline,substring(1,#gI-2,gI));
+    inStr:=l1|newline|l2|newline|Igens;
+    mIn<<inStr<<close;
+    mOut:=temporaryFileName()|".ms";
+    if opt#Verbosity>0 then<< "msolve input file is called: " << mIn << endl;
+    if opt#Verbosity>0 then<< "msolve output file is called: " << mOut << endl;
+    callStr:=msolveEXEwithOpts(opt)|" -e "|toString(elimNum)|" -g 2 -f "|mIn|" -o "|mOut;
+    run(callStr);
+    if char R === 0 then 
+      ideal readMsolveOutputFile(elimR, mOut)
+    else
+      ideal readMsolveOutputFile(R, mOut)
```

Comment:
What is this for?

Reply:
This is necessary. 

msolve's elimination has different behavior for QQ vs a finite field.  

Over QQ only the generators of the elimination ideal in the smaller ring are returned (they don't want to do rational reconstruction on the generators you don't need)

But over a finite field they return the full GB 

See the docs in Msolve.m2 for 	msolveEliminate